### ISSUE TITLE
Question on Bundling Plugins with the Theme

### ISSUE THREAD

**Issue by Radizzt** [wprig/wprig#685]
## Issue Overview

I Want to bundle [Advanced Custom Field](https://www.advancedcustomfields.com/) with my theme and I'm wondering whats the best practice for it with WP Rig?
## Describe your environment

Local by Flywheel (Version 5.9.9+5246)
VSCode
Chrome
## Possible Solution

The plugin itself has [documentation](https://www.advancedcustomfields.com/resources/including-acf-within-a-plugin-or-theme/) on how to do it, but since WP Rig modernize normal WordPress I'm curious if I should follow their documentation, or do you suggest something else?

Thanks in advance.

**Comment by heathernew09** [comment]
I've been experimenting with this. Right now I've implemented by these steps. 

1. Create a new folder in themes/wprig/inc and name it ACF
2. Inside the ACF folder create a php file named Component.php (capitalize)
Setup the component class as shown in the code below and modify to your preference
`<?php
/**
 * WP_Rig\WP_Rig\ACF\Component class
 *
 * @package wp_rig
 */

namespace WP_Rig\WP_Rig\ACF;

use WP_Rig\WP_Rig\Component_Interface;
use function add_action;

/**
 * Class for implementing ACF theme options support.
 */
class Component implements Component_Interface {

	/**
	 * Gets the unique identifier for the theme component.
	 *
	 * @return string Component slug.
	 */
	public function get_slug() : string {
		return 'acf';
	}

	/**
	 * Adds the action and filter hooks to integrate with WordPress.
	 */
	public function initialize() {
		add_action( 'after_setup_theme', array( $this, 'add_acf_theme_options' ), 10, 2 );
	}
	/**
	 *  Adds ACF Options Menu
	 * */
	public function add_acf_theme_options() {
		if ( function_exists( 'acf_add_options_page' ) ) {
			acf_add_options_page();
		}
	}
	/**
	 * Register acf fields to WordPress API
	 *
	 * @param string $response Response of ACF to Rest.
	 * @param string $post Grabs the Posts from API.
	 * @param string $request Grabs the content we want from API.
	 * @link https://support.advancedcustomfields.com/forums/topic/json-rest-api-and-acf/
	 */
	public function acf_to_rest_api( $response, $post, $request ) {
		if ( ! function_exists( 'get_fields' ) ) return $response;

		if ( isset( $post ) ) {
			$acf = get_fields( $post->id );
			$response->data['acf'] = $acf;
		}
		return $response;
	}
}`
3. In theme folder open Theme.php
	Locate get_default_components and add ACF to the array
	`new ACF\Component(),`

**Comment by heathernew09** [comment]
I am also utilizing acf-json so you may want to give that a go and add that folder to your theme as well. https://www.advancedcustomfields.com/resources/local-json/

**Comment by Radizzt** [comment]
Hey @heathernew0,
Thanks for the info, but can you give me a bit more info on how this works?
This is a bit over my head.
I generally create a field group from the dashboard and then put it into my template files.

I'm assuming you have the ACF plugin in the _wp_content/plugin/_ folder and activated it because without activating it the  option you created through `acf_add_options_page()` wont work

I added the Component.php in _inc/_ folder and register the Component in Theme.php.
I also created the _acf-json/_ folder right under the _theme/_ folder as well.

I'm not 100% sure what does the `acf_to_rest_api()` function does, or how I would use it. Do you mind giving me a bit more detail on it?

For simplicity sake, I made a simple page template just to test ACF as if I install it through dashboard
```php
<?php
/*
Template Name: ACF Test
*/
get_header(); ?>
	<main id="primary" class="site-main">
	<h1>homepage</h1>
		<div><?php the_field('text'); ?></div>
		<?php
			$image = get_field('image');
			if( !empty( $image ) ): ?>
				<img src="<?php echo esc_url($image['url']); ?>" alt="<?php echo esc_attr($image['alt']); ?>" />
		<?php endif; ?>
	</main><!-- #primary -->
<?php
get_footer();
```

How would I code it differently with the new Component class?

Thanks in advance

**Comment by dthenley** [comment]
@Radizzt you wouldn't code it any differently than what you have there. The component class @heathernew09 provided is only to include ACF theme options into your theme which is extremely useful for creating and accessing global variables, you can read more about that here https://www.advancedcustomfields.com/resources/options-page/.

As for the act_to_rest_api that is mainly used when accessing the wordpress rest api. This is used a lot for headless wordpress applications.

**Comment by Radizzt** [comment]
@dthenley Thanks for clearing thing ups for me. 
I manage to set it to a point where they would need to download the plugin and create the fields themselves. Correct me if I'm wrong, I don't think it's possible to create a field group within a theme as it requires the database.

My only issue is if I plan to put a repeater or any ACF Pro option in there. It won't be possible, but I think this is a separate issue. For now, I'm just pulling info from posts and give it a repeater-esque functionality

**Comment by dthenley** [comment]
If you  want to include acf fields into your theme you can do this by creating the acf-json folder in your theme root folder. This will add any custom fields from your theme and make it so that all the user would need to do is sync to edit but they should be ready for immediate use upon installation. When you do this you will want to edit the export section in the `config/config.json`  to add a modified version of this 

```
{
  "dev": {
  "export": {
    "compress": true,
    "generatePotFile": true,
    "filesToCopy": [
      "LICENSE",
      "readme.txt",
      "screenshot.png",
      "assets/css/vendor/**/*.css",
      "assets/js/vendor/**/*.js",
     "acf-json/*"
    ]
  }
}
```
As for a way to make a plugin required when installing a theme i found the best way to do that is with https://github.com/TGMPA/TGM-Plugin-Activation.

**Comment by Radizzt** [comment]
@dthenley Sorry for reopening an old thread, but my theme is finally ready to bundle and my issue now is the .json files that were generated in the `acf-json/` folder are not being copied over when I bundle it. The `index.php` I created within the `acf-json/` folder is being copied over, but the `.json` file are not. 

I added an extra line `"acf-json/*.json"` just in case it wasn't taking in JSON file, but no luck so far.

```
"export": {
      "compress": true,
      "generatePotFile": true,
      "filesToCopy": [
        "LICENSE",
        "readme.txt",
        "screenshot.png",
        "assets/css/vendor/**/*.css",
        "assets/js/vendor/**/*.js",
        "acf-json/*",
        "acf-json/*.json"
      ]
    }
```

Any idea why this is happening?
